File: src/ui/layout.ts
/**
 * Jetons de design et mise en page de l'écran de combat (docs/UX.md, tickets
 * E04-T02 et E04-T03). Tout est exprimé en pixels LOGIQUES sur une toile de
 * 480×854 (portrait) ; le rendu est mis à l'échelle de l'écran par
 * `renderScale`. Aucun import de Phaser : le fichier est testé sans navigateur.
 */

export const GAME_W = 480;
export const GAME_H = 854;

/** Cible tactile minimale (px logiques) : règle UX « ≥ 48 px ». */
export const MIN_TOUCH = 48;

/** Tailles de texte (px logiques). Aucune n'est inférieure à `FONT.small` (règle UX « ≥ 14 px »). */
export const FONT = {
  small: 14,
  body: 16,
  strong: 18,
  title: 20,
  banner: 28,
  cooldown: 24,
} as const;

/** Marges de sécurité : encoches et barres système (à vérifier sur appareil réel, E11-T01). */
export const SAFE = { top: 8, bottom: 40, side: 12 } as const;

export const GAP = 8;

export const COLOR = {
  background: 0x10121a,
  panel: 0x2a2d3a,
  panelStroke: 0x555a70,
  text: "#ffffff",
  textMuted: "#a7adc2",
  heal: "#7CFFB2",
  damage: "#FF7C7C",
  poison: "#C78CFF",
  shield: "#7CC8FF",
  boss: "#E6E6E6",
  danger: "#FF5B5B",
  selected: 0xffe066,
  hpHigh: 0x5fd35f,
  hpMid: 0xf0a23a,
  hpLow: 0xe0443e,
  mana: 0x4aa8ff,
  role: { tank: 0x4a6fa5, dps: 0xb5495b, healer: 0x4caf7d } as Record<string, number>,
} as const;

export interface Rect {
  x: number;
  y: number;
  w: number;
  h: number;
}

/** Zones verticales de l'écran, de haut en bas (voir docs/UX.md §4). */
export const ZONES = {
  topBar: { x: 0, y: SAFE.top, w: GAME_W, h: 52 },
  boss: { x: 0, y: 64, w: GAME_W, h: 186 },
  band: { x: SAFE.side, y: 262, w: GAME_W - 2 * SAFE.side, h: 66 },
  team: { x: SAFE.side, y: 340, w: GAME_W - 2 * SAFE.side, h: 200 },
  strip: { x: SAFE.side, y: 552, w: GAME_W - 2 * SAFE.side, h: 64 },
  skills: { x: SAFE.side, y: 632, w: GAME_W - 2 * SAFE.side, h: 164 },
} satisfies Record<string, Rect>;

/** Bouton de pause : cible tactile ≥ 48 px, en haut à droite. */
export const PAUSE_BUTTON: Rect = { x: GAME_W - SAFE.side - MIN_TOUCH, y: SAFE.top, w: MIN_TOUCH, h: MIN_TOUCH };

/** Barre de PV du boss : s'arrête avant le bouton de pause pour ne pas le chevaucher. */
export const BOSS_HP_BAR: Rect = {
  x: SAFE.side,
  y: SAFE.top + 34,
  w: PAUSE_BUTTON.x - GAP - SAFE.side,
  h: 14,
};

/** Répartit `count` éléments égaux sur la largeur d'une zone, avec `GAP` entre eux. */
export function splitRow(zone: Rect, count: number): Rect[] {
  const width = (zone.w - GAP * (count - 1)) / count;
  return Array.from({ length: count }, (_, i) => ({
    x: zone.x + i * (width + GAP),
    y: zone.y,
    w: width,
    h: zone.h,
  }));
}

export const teamCardRects = (count: number): Rect[] => splitRow(ZONES.team, count);
export const skillButtonRects = (count: number): Rect[] => splitRow(ZONES.skills, count);

/** Couleur de la barre de PV selon le ratio (seuils lisibles ; la couleur n'est jamais seule, le chiffre est affiché). */
export function hpColor(ratio: number): number {
  if (ratio > 0.6) return COLOR.hpHigh;
  if (ratio > 0.3) return COLOR.hpMid;
  return COLOR.hpLow;
}

/** Borne le ratio de pixels de l'appareil : net sur écrans denses, sans surcoût excessif. */
export function clampRenderScale(devicePixelRatio: number): number {
  if (!Number.isFinite(devicePixelRatio) || devicePixelRatio < 1) return 1;
  return Math.min(Math.ceil(devicePixelRatio), 3);
}

/**
 * Échelle de rendu de l'appareil. En développement, `?dpr=2` dans l'adresse
 * force une échelle pour vérifier le rendu haute densité sur un écran normal.
 */
export function renderScale(): number {
  if (typeof window === "undefined") return 1;
  if (import.meta.env.DEV) {
    const forced = Number(new URLSearchParams(window.location.search).get("dpr"));
    if (forced >= 1) return clampRenderScale(forced);
  }
  return clampRenderScale(window.devicePixelRatio);
}

File: core/Healer.Ui/Layout.cs
using System;
using System.Collections.Generic;

namespace Healer.Ui
{
    /// <summary>Rectangle en pixels logiques de l'écran de combat (origine en haut à gauche).</summary>
    public readonly struct Rect
    {
        public double X { get; }
        public double Y { get; }
        public double W { get; }
        public double H { get; }

        public Rect(double x, double y, double w, double h)
        {
            X = x; Y = y; W = w; H = h;
        }

        public double Right => X + W;
        public double Bottom => Y + H;

        public bool Overlaps(Rect o) => X < o.Right && o.X < Right && Y < o.Bottom && o.Y < Bottom;
        public bool InsideScreen() => X >= 0 && Y >= 0 && Right <= Layout.GameW && Bottom <= Layout.GameH;
    }

    /// <summary>Bande de couleur de la barre de PV (la couleur n'est jamais seule : la valeur est toujours affichée).</summary>
    public enum HpBand
    {
        High,
        Mid,
        Low,
    }

    /// <summary>
    /// Jetons de design et mise en page PAYSAGE de l'écran de combat (décision D-042 : PC d'abord, mobile
    /// toujours compatible ; docs/UX.md, tickets E04-T02 et E04-T03), en pixels LOGIQUES sur une toile de
    /// 1280×720 (16:9). Le client Unity adapte cette grille à l'écran ; les règles UX (cibles ≥ 48 px, texte
    /// ≥ 14 px, marges de sécurité) sont testées ici sans Unity. La version Phaser gardait la grille portrait 480×854.
    /// </summary>
    public static class Layout
    {
        public const double GameW = 1280;
        public const double GameH = 720;

        /// <summary>Cible tactile minimale : règle UX « ≥ 48 px ».</summary>
        public const double MinTouch = 48;

        /// <summary>Tailles de texte (px logiques). Aucune n'est inférieure à Small (règle UX « ≥ 14 px »).</summary>
        public static class Font
        {
            public const int Small = 14;
            public const int Body = 16;
            public const int Strong = 18;
            public const int Title = 20;
            public const int Banner = 28;
            public const int Cooldown = 24;

            public static readonly IReadOnlyDictionary<string, int> All = new Dictionary<string, int>
            {
                ["small"] = Small, ["body"] = Body, ["strong"] = Strong,
                ["title"] = Title, ["banner"] = Banner, ["cooldown"] = Cooldown,
            };
        }

        /// <summary>Marges de sécurité (encoches, barres système) ; à vérifier sur appareil réel (E11-T01).</summary>
        public const double SafeTop = 8;
        public const double SafeBottom = 16;
        public const double SafeSide = 12;
        public const double Gap = 8;

        /// <summary>Zones verticales de l'écran, de haut en bas (docs/UX.md §4).</summary>
        public static class Zones
        {
            /// <summary>Colonne centrale de 1000 px : cartes et sorts y sont regroupés (lecture du regard sans balayer 1280 px).</summary>
            public const double CenterX = 140;
            public const double CenterW = 1000;

            public static readonly Rect TopBar = new Rect(0, SafeTop, GameW, 52);
            public static readonly Rect Boss = new Rect(0, 64, GameW, 250);
            public static readonly Rect Band = new Rect(CenterX, 322, CenterW, 44);
            public static readonly Rect Team = new Rect(CenterX, 388, CenterW, 136);
            public static readonly Rect Strip = new Rect(CenterX, 534, CenterW, 48);
            public static readonly Rect Skills = new Rect(CenterX, 592, CenterW, 104);

            public static IReadOnlyList<(string name, Rect rect)> InOrder => new[]
            {
                ("topBar", TopBar), ("boss", Boss), ("band", Band), ("team", Team), ("strip", Strip), ("skills", Skills),
            };
        }

        /// <summary>Bouton de pause : cible tactile ≥ 48 px, en haut à droite.</summary>
        public static readonly Rect PauseButton = new Rect(GameW - SafeSide - MinTouch, SafeTop, MinTouch, MinTouch);

        /// <summary>Boutons des écrans modaux (démarrage, bilan). Volontairement au-dessus de la mise en page : voir InputGate.</summary>
        public static readonly Rect StartButton = new Rect(GameW / 2 - 120, 470, 240, 60);
        public static readonly Rect RestartButton = new Rect(GameW / 2 - 120, 450, 240, 56);
        /// <summary>Barre de PV du boss : s'arrête avant le bouton de pause pour ne pas le chevaucher.</summary>
        public static readonly Rect BossHpBar = new Rect(Zones.CenterX + 100, SafeTop + 34, 800, 14);

        /// <summary>Répartit `count` éléments égaux sur la largeur d'une zone, avec Gap entre eux.</summary>
        public static Rect[] SplitRow(Rect zone, int count)
        {
            double width = (zone.W - Gap * (count - 1)) / count;
            var result = new Rect[count];
            for (int i = 0; i < count; i++) result[i] = new Rect(zone.X + i * (width + Gap), zone.Y, width, zone.H);
            return result;
        }

        public static Rect[] TeamCardRects(int count) => SplitRow(Zones.Team, count);
        public static Rect[] SkillButtonRects(int count) => SplitRow(Zones.Skills, count);

        /// <summary>Bande de la barre de PV selon le ratio (seuils lisibles : > 60 %, > 30 %, sinon bas).</summary>
        public static HpBand HpBandFor(double ratio)
        {
            if (ratio > 0.6) return HpBand.High;
            if (ratio > 0.3) return HpBand.Mid;
            return HpBand.Low;
        }

        /// <summary>Borne le ratio de pixels de l'appareil : net sur écrans denses, sans surcoût excessif.</summary>
        public static int ClampRenderScale(double devicePixelRatio)
        {
            if (double.IsNaN(devicePixelRatio) || double.IsInfinity(devicePixelRatio) || devicePixelRatio < 1) return 1;
            return (int)Math.Min(Math.Ceiling(devicePixelRatio), 3);
        }
    }
}
